### ISSUE TITLE
Bug Report + Proposal: Object.assign()

### ISSUE THREAD

**Issue by Benjamin-Dobell** [microsoft/TypeScript#31982]
**TypeScript Version:**  3.4.0-dev.201xxxxx

**Search Terms:** Object.assign return type

**Code**

1.

```ts
let test1: string = Object.assign({ a: 1 }, { a: "one" }).a
// $ExpectError
let test2: number = Object.assign({ a: 1 }, { a: "one" }).a // No error
```

2.
```ts
// $ExpectError
Object.assign({ a: 1, b: 2 }, "hi").toLowerCase() // No error
```

3.

```ts
// $ExpectError
Object.assign({ a: 1, b: 2 }, "hi").length // No error
```

**Expected behavior:**

1.

> Type 'string' is not assignable to type 'number'

2.

> Property 'toLowerCase' does not exist on type ...

3. 

> Property 'length' does not exist on type ...

**Actual behavior:**

No errors raised.

**Playground Link:**  [Link](https://www.typescriptlang.org/play/#src=%2F%2F%20Pass%0D%0Alet%20test1%3A%20string%20%3D%20Object.assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20%22one%22%20%7D).a%0D%0A%0D%0A%2F%2F%20Failure%0D%0A%2F%2F%20Should%20raise%3A%0D%0A%2F%2F%20%20%20%20%20Type%20'string'%20is%20not%20assignable%20to%20type%20'number'%0D%0A%2F%2F%20%24ExpectError%0D%0Alet%20test2%3A%20number%20%3D%20Object.assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20%22one%22%20%7D).a%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test3%3A%20string%20%3D%20Object.assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20(0%20as%20any)%20%7D).a%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test4%3A%20number%20%3D%20Object.assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20undefined%20%7D).a%0D%0A%0D%0A%2F%2F%20Pass...%20sort%20of%0D%0A%2F%2F%20Should%20raise%3A%0D%0A%2F%2F%20%20%20%20%20Type%20'number'%20is%20not%20assignable%20to%20type%20'string'%0D%0A%2F%2F%20Note%3A%20The%20the%20error%20message%20is%20incorrect%20and%20demonstrative%20of%20the%20issue%20at%20hand.%0D%0A%2F%2F%20%24ExpectError%0D%0Alet%20test5%3A%20string%20%3D%20Object.assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20%7B%7D%20%7D).a%20%0D%0A%0D%0A%2F%2F%20Failure%0D%0A%2F%2F%20Should%20raise%3A%0D%0A%2F%2F%20%20%20%20%20Property%20'toLowerCase'%20does%20not%20exist%20on%20type%20...%0D%0A%2F%2F%20%24ExpectError%0D%0AObject.assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22).toLowerCase()%20%2F%2F%20test6%0D%0A%0D%0A%2F%2F%20Pass%0D%0AObject.assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22)%5B%22a%22%5D%20%2F%2F%20test7%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test8%3A%20string%20%3D%20Object.assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22)%5B0%5D%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Type%20'string'%20is%20not%20assignable%20to%20type%20'number'%0D%0A%2F%2F%20%24ExpectError%0D%0Alet%20test9%3A%20number%20%3D%20Object.assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22)%5B0%5D%0D%0A%0D%0A%2F%2F%20Failure%0D%0A%2F%2F%20Should%20raise%3A%0D%0A%2F%2F%20%20%20%20%20Property%20'length'%20does%20not%20exist%20on%20type%20...%0D%0A%2F%2F%20%24ExpectError%0D%0AObject.assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22).length%20%2F%2F%20test10%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test11%3A%20number%20%3D%20Object.assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%5B1%2C%202%2C%203%5D)%5B0%5D%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Type%20'number'%20is%20not%20assignable%20to%20type%20'string'%0D%0A%2F%2F%20%24ExpectError%0D%0Alet%20test12%3A%20string%20%3D%20Object.assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%5B1%2C%202%2C%203%5D)%5B0%5D%0D%0A)

**Related Issues:** 

* https://github.com/microsoft/TypeScript/pull/30481
* https://github.com/microsoft/TypeScript/issues/29348

# Proposal

```ts
type IndexerReturnType<T extends ArrayLike<any>> = T extends ArrayLike<infer V> ? V : any;

type ArrayLikeIndexer<T> = T extends ArrayLike<any> ? { [k: number]: IndexerReturnType<T> } : never 

type PickDefined<T> = {
    [P in keyof T]: T[P] extends undefined ? never : T[P]
};

type Assign<A, B> = A extends object ? (
    B extends object ?
        Pick<A, Exclude<keyof A, keyof B>> & PickDefined<B> :
        A & ArrayLikeIndexer<B>
) : (A & B)

declare function assign<A, B>(a: A, b: B): Assign<A, B>
```

**Note 1**: _`PickDefined<T>` is necessary with `strictNullChecks` enabled._
**Note 2**: _In the definition of `Assign<A, B>` I would have expected the usage of `keyof B` to need to be replaced with `keyof PickDefined<B>`. It seems, that's _presently_ not the case - so I've opted to omit it and avoid any potential slow-down. Nonetheless, I must admit that I'm hazy as to why this is permissible._

Obviously to keep things brief I've just covered the single `source` use-case rather than providing overloads with several source. The definition is nonetheless chainable i.e. `Assign<A, Assign<B, C>>` etc. 

**Playground Link:** [Link](https://www.typescriptlang.org/play/#src=%2F%2F%20*****%20Proposed%20Definition%20*****%0D%0A%0D%0Atype%20IndexerReturnType%3CT%20extends%20ArrayLike%3Cany%3E%3E%20%3D%20T%20extends%20ArrayLike%3Cinfer%20V%3E%20%3F%20V%20%3A%20any%3B%0D%0A%0D%0Atype%20ArrayLikeIndexer%3CT%3E%20%3D%20T%20extends%20ArrayLike%3Cany%3E%20%3F%20%7B%20%5Bk%3A%20number%5D%3A%20IndexerReturnType%3CT%3E%20%7D%20%3A%20never%20%0D%0A%0D%0Atype%20PickDefined%3CT%3E%20%3D%20%7B%0D%0A%20%20%20%20%5BP%20in%20keyof%20T%5D%3A%20T%5BP%5D%20extends%20undefined%20%3F%20never%20%3A%20T%5BP%5D%0D%0A%7D%3B%0D%0A%0D%0Atype%20Assign%3CA%2C%20B%3E%20%3D%20A%20extends%20object%20%3F%20(%0D%0A%20%20%20%20B%20extends%20object%20%3F%20Pick%3CA%2C%20Exclude%3Ckeyof%20A%2C%20keyof%20B%3E%3E%20%26%20PickDefined%3CB%3E%20%3A%20A%20%26%20ArrayLikeIndexer%3CB%3E%0D%0A)%20%3A%20(A%20%26%20B)%0D%0A%0D%0A%2F%2F%20Note%3A%20In%20the%20above%20I%20would%20have%20expected%20%60keyof%20B%60%20to%20need%20to%20be%20%60keyof%20PickDefined%3CB%3E%60%2C%0D%0A%2F%2F%20however%20that's%20*presently*%20not%20the%20case%20-%20so%20no%20point%20slowing%20TypeScript%20down.%0D%0A%0D%0Adeclare%20function%20assign%3CA%2C%20B%3E(a%3A%20A%2C%20b%3A%20B)%3A%20Assign%3CA%2C%20B%3E%0D%0A%0D%0A%2F%2F%20*****%20Tests%20*****%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test1%3A%20string%20%3D%20assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20%22one%22%20%7D).a%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Type%20'string'%20is%20not%20assignable%20to%20type%20'number'%0D%0A%2F%2F%20%24ExpectError%0D%0Alet%20test2%3A%20number%20%3D%20assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20%22one%22%20%7D).a%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test3%3A%20string%20%3D%20assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20(0%20as%20any)%20%7D).a%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test4%3A%20number%20%3D%20assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20undefined%20%7D).a%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Type%20'number'%20is%20not%20assignable%20to%20type%20'string'%0D%0A%2F%2F%20%24ExpectError%0D%0Alet%20test5%3A%20string%20%3D%20assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20%7B%7D%20%7D).a%20%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Property%20'toLowerCase'%20does%20not%20exist%20on%20type%20...%0D%0A%2F%2F%20%24ExpectError%0D%0Aassign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22).toLowerCase()%20%2F%2F%20test6%0D%0A%0D%0A%2F%2F%20Pass%0D%0Aassign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22)%5B%22a%22%5D%20%2F%2F%20test7%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test8%3A%20string%20%3D%20assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22)%5B0%5D%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Type%20'string'%20is%20not%20assignable%20to%20type%20'number'%0D%0A%2F%2F%20%24ExpectError%0D%0Alet%20test9%3A%20number%20%3D%20assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22)%5B0%5D%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Property%20'length'%20does%20not%20exist%20on%20type%20...%0D%0A%2F%2F%20%24ExpectError%0D%0Aassign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22).length%20%2F%2F%20test10%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test11%3A%20number%20%3D%20assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%5B1%2C%202%2C%203%5D)%5B0%5D%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Type%20'number'%20is%20not%20assignable%20to%20type%20'string'%0D%0A%2F%2F%20%24ExpectError%0D%0Alet%20test12%3A%20string%20%3D%20assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%5B1%2C%202%2C%203%5D)%5B0%5D%0D%0A) (For posterity please enable `strictNullChecks`)

**Comment by jcalz** [comment]
Relevant [comment](https://github.com/microsoft/TypeScript/pull/28553#issuecomment-440004598) (in #28553) saying why this is not already being done:
>  Spreading two overlapping types is pretty rare, so we decided not to create a type operator for it.

Relevant [possible implementation](https://github.com/Microsoft/TypeScript/pull/21316#issuecomment-359574388) (in #21316) of a `Spread<L, R>` which is similar to `Assign<A, B>`.   I don't know if anyone has a demonstrably superior implementation of this thing given all the potential edge cases like unknown, optional, and non-own properties:

```ts
class A {
  c?: boolean;
  constructor(public a: string) {}
  sayA() {
    console.log(this.a);
  }
}
class B extends A {
  constructor(a: string, public b: string) {
    super(a);
  }
}
function hmm(a: A) {
  return assign({ b: 123, c: 456 }, a);
}
const result = hmm(new B("a", "b"));

// unknown properties
result.b; // number at compile time, but:
console.log(typeof result.b); // "string"

// optional properties
result.c; // boolean | undefined at compile time, but:
console.log(typeof result.c); // "number"

// non-own properties
result.sayA; // function at compile time, but:
console.log(typeof result.sayA); // "undefined"
```
[Link to code](https://typescript-play.js.org/#code/C4TwDgpgBAkgdgEwgDwgJwEoWAVzXAFXAgB4CoVgJEBnKAQTTQEMQAZASwGtTm4QAfAKgBeKOUrUEdRi3bdSHOADN0UAGoCAUFCgB+DTqgAuKHxABuLVtCQGTVpx7wkqNGWFiJyKrXtynXn5tXQMAbygAbS5TOBwAWwAjdABdUxcUdCxcfCJIDygAXyNYiAA3dCsbYigABQ4AYy4AEQhlJQgEArEIyNqoJSgeEAB7ZXE08T6Uih8pOhxENo6EfSg4crVTAmmiqttoehoaDgBzOBJ6ABooACFPBlnfaSgRxIArCAbgIwMACluT3mrw+Xx+ugha3qTUuNwAosgGgAbHBIEjDMYMG4Y8b3YQAMjqjRayw2XXuRkhpnoUEJskcCgybhI9wAlCUoH8aYTbqyqspFt8OCM4GZjmcLtc7gI-sxqTdEqZedTxedYdKoGFKVA0Ng8KKAPKg74AOmYqrgsoVrLFZn4VmKWmR5pkmqMDT0pkSIxGSIgfCsugaIpowDQOG+IzQfzAOESSMaZlMobQSlONrCxV0NFY9D+Ge1wbgNF9EBNSJGpz+wAAFhwaGa+UZisVncc7kC-DStUGQ2GI8Ao7Lk2G0zdY-HE4qoCm0wXITQcJBo8wm7oW1oBXAhSKoDX4vFhwx5zq9fgxSdzn8ItOAIwAJgAzDcGqYACwAVgAbEUbquHU6fZQIkojrBAADudx-AARMw0E3NBiTQU2gHFsAp6Lki6FiPuh6JE2uqYcAJqJBYUAAPTkesCTJGgZjocG8RgBwfpQMAHDxBACo4MAxioSWfrlpW1bEJihE4FhJF8hRVHQbOcCnNB1jiZJDRkZRwE+n6fBQAAPlAixIO0ZL0VAjHMax7GcdxvH8aWQlVgcYkQERJoNNJGnQXESToEpWgqcROYgPQ6lUVuO6iswDEjExLHQFZXHATxfFFgJZYVo5onjAFJpBfQHmyYZpKdNBQA)

**Comment by Benjamin-Dobell** [comment]
Thanks, @jcalz, and great example!

> Spreading two overlapping types is pretty rare, so we decided not to create a type operator for it.

It's bit of a shame as there's been a _heap_ of improvements in soundness recently, really loving the direction!

Not that I think it's particularly compelling on its lonesome, but just for reference, I am seeing overlapping types being assigned. Specifically custom "sub-classes", default props of `number[]` being overwritten with `string[]`, `Date[]` etc. Although, in the aforementioned case it's actually with a custom _recursive_ assign in a JS library ([chartist](https://github.com/gionkunz/chartist-js/blob/develop/src/scripts/core.js#L61)). I was improving the DefinitelyTyped definitions for chartist when I noticed `Object.assign` _seemingly_ wasn't up to scratch. 

Anyway, I do totally understand the decision though. I guess it'd only be feasible with first class support for "exact types" (https://github.com/Microsoft/TypeScript/issues/12936), and my head is hurting just thinking about how much that'd further complicate the types here.

I also noted that when chained my `Assign` isn't propagating the indexer; technically it needs to create a new indexer with a union of the return types in cases of:

```ts
let x: number
Object.assign({}, "abcde", [1, 2, 3])[x] // string | number
```

Hopefully this will become a bit more feasible in near future; I'm a sucker for soundness.

**Comment by Benjamin-Dobell** [comment]
For future reference, whilst without a doubt it's far superior to my own attempt, @ahejlsberg's implementation unfortunately does have a couple of failures when [run against my test suite](https://www.typescriptlang.org/play/#src=type%20Diff%3CT%2C%20U%3E%20%3D%20T%20extends%20U%20%3F%20never%20%3A%20T%3B%20%0D%0A%0D%0A%2F%2F%20Names%20of%20properties%20in%20T%20with%20types%20that%20include%20undefined%0D%0Atype%20OptionalPropertyNames%3CT%3E%20%3D%0D%0A%20%20%20%20%7B%20%5BK%20in%20keyof%20T%5D%3A%20undefined%20extends%20T%5BK%5D%20%3F%20K%20%3A%20never%20%7D%5Bkeyof%20T%5D%3B%0D%0A%0D%0A%2F%2F%20Common%20properties%20from%20L%20and%20R%20with%20undefined%20in%20R%5BK%5D%20replaced%20by%20type%20in%20L%5BK%5D%0D%0Atype%20SpreadProperties%3CL%2C%20R%2C%20K%20extends%20keyof%20L%20%26%20keyof%20R%3E%20%3D%0D%0A%20%20%20%20%7B%20%5BP%20in%20K%5D%3A%20L%5BP%5D%20%7C%20Diff%3CR%5BP%5D%2C%20undefined%3E%20%7D%3B%0D%0A%0D%0A%2F%2F%20Type%20of%20%7B%20...L%2C%20...R%20%7D%0D%0Atype%20Spread%3CL%2C%20R%3E%20%3D%0D%0A%20%20%20%20%2F%2F%20Properties%20in%20L%20that%20don't%20exist%20in%20R%0D%0A%20%20%20%20%26%20Pick%3CL%2C%20Diff%3Ckeyof%20L%2C%20keyof%20R%3E%3E%0D%0A%20%20%20%20%2F%2F%20Properties%20in%20R%20with%20types%20that%20exclude%20undefined%0D%0A%20%20%20%20%26%20Pick%3CR%2C%20Diff%3Ckeyof%20R%2C%20OptionalPropertyNames%3CR%3E%3E%3E%0D%0A%20%20%20%20%2F%2F%20Properties%20in%20R%2C%20with%20types%20that%20include%20undefined%2C%20that%20don't%20exist%20in%20L%0D%0A%20%20%20%20%26%20Pick%3CR%2C%20Diff%3COptionalPropertyNames%3CR%3E%2C%20keyof%20L%3E%3E%0D%0A%20%20%20%20%2F%2F%20Properties%20in%20R%2C%20with%20types%20that%20include%20undefined%2C%20that%20exist%20in%20L%0D%0A%20%20%20%20%26%20SpreadProperties%3CL%2C%20R%2C%20OptionalPropertyNames%3CR%3E%20%26%20keyof%20L%3E%3B%0D%0A%0D%0Adeclare%20function%20assign%3CA%2C%20B%3E(a%3A%20A%2C%20b%3A%20B)%3A%20Spread%3CA%2C%20B%3E%0D%0A%0D%0A%2F%2F%20*****%20Tests%20*****%0D%0A%0D%0A%2F%2F%20Failure%0D%0Alet%20test1%3A%20string%20%3D%20assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20%22one%22%20%7D).a%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Type%20'string'%20is%20not%20assignable%20to%20type%20'number'%0D%0A%2F%2F%20%24ExpectError%0D%0Alet%20test2%3A%20number%20%3D%20assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20%22one%22%20%7D).a%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test3%3A%20string%20%3D%20assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20(0%20as%20any)%20%7D).a%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test4%3A%20number%20%3D%20assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20undefined%20%7D).a%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Type%20'number'%20is%20not%20assignable%20to%20type%20'string'%0D%0A%2F%2F%20%24ExpectError%0D%0Alet%20test5%3A%20string%20%3D%20assign(%7B%20a%3A%201%20%7D%2C%20%7B%20a%3A%20%7B%7D%20%7D).a%20%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Property%20'toLowerCase'%20does%20not%20exist%20on%20type%20...%0D%0A%2F%2F%20%24ExpectError%0D%0Aassign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22).toLowerCase()%20%2F%2F%20test6%0D%0A%0D%0A%2F%2F%20Pass%0D%0Aassign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22)%5B%22a%22%5D%20%2F%2F%20test7%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test8%3A%20string%20%3D%20assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22)%5B0%5D%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Type%20'string'%20is%20not%20assignable%20to%20type%20'number'%0D%0A%2F%2F%20%24ExpectError%0D%0Alet%20test9%3A%20number%20%3D%20assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22)%5B0%5D%0D%0A%0D%0A%2F%2F%20Failure%0D%0A%2F%2F%20Should%20raise%3A%0D%0A%2F%2F%20%20%20%20%20Property%20'length'%20does%20not%20exist%20on%20type%20...%0D%0A%2F%2F%20%24ExpectError%0D%0Aassign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%22hi%22).length%20%2F%2F%20test10%0D%0A%0D%0A%2F%2F%20Pass%0D%0Alet%20test11%3A%20number%20%3D%20assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%5B1%2C%202%2C%203%5D)%5B0%5D%0D%0A%0D%0A%2F%2F%20Pass%0D%0A%2F%2F%20Correctly%20raises%3A%0D%0A%2F%2F%20%20%20%20%20Type%20'number'%20is%20not%20assignable%20to%20type%20'string'%0D%0A%2F%2F%20%24ExpectError%0D%0Alet%20test12%3A%20string%20%3D%20assign(%7B%20a%3A%201%2C%20b%3A%202%20%7D%2C%20%5B1%2C%202%2C%203%5D)%5B0%5D%0D%0A). Failing `test1` in particular seems non-ideal. However, just to be clear, @ahejlsberg does a _much_ better job of dealing with prickly edge cases (i.e. those mentioned above).

**Comment by jcalz** [comment]
You've got `--strictNullChecks` disabled there (well, the Playground does that by default), and therefore the inferred type of `{a: "one"}` is `{a: string}` which is also satisfied by `{a: undefined}`.  And since that implementation of `Spread<L, R>` treats any possibly-undefined property as optional, you get that behavior.  It is possible to detect optional properties in a way that distinguishes them from required-but-possibly-undefined properties:

```ts
type OptionalPropertyNames<T> = { [K in keyof T]-?:
  ({} extends { [P in K]: T[K] } ? K : never)
}[keyof T]
```

which should fix that particular issue, but I feel unclean even thinking about what happens when `--strictNullChecks` is disabled.   Even with `--strictNullChecks` enabled it's still not really possible to consistently distinguish [missing from `undefined`](https://github.com/Microsoft/TypeScript/issues/13195) in TypeScript, which makes it hard to even say what the "right" thing to do is.  If your test suite contains `assign({a: 1}, {a: undefined})`, you are probably heading down a dark path and need to turn back before it is too late.

**Comment by Benjamin-Dobell** [comment]
> You've got `--strictNullChecks` disabled there

Ugh, spot on, thank you! 😊

... and that's my call to go to bed (don't look at my time zone :wink:).


> If your test suite contains `assign({a: 1}, {a: undefined})`, you are probably heading down a dark path and need to turn back before it is too late.

Looks like I'm too far gone. 😉 

However, in all seriousness, @ahejlsberg's implementation is fantastic and ticks all the boxes, my sincerest apologies for falsely calling out failing tests!

I'm not sure whether this issue should remain open to track the state of `Object.assign`, however if the TypeScript team wish to, I'd have little objections to this being closed.

**Comment by RyanCavanaugh** [comment]
There's an additional trade-off here because it's desirable for `Object.assign(T, U)` for two type parameters `T` and `U` to be assignable to both `T` and `U` (this comes up a *lot* in various code examples we've gotten reports of) - but in reality this *isn't* actually knowable unless you can prove that `T` and `U` will be instantiated with compatible-or-disjoint properties. Same for `Object.assign(e, T)`; it is generally agreed that this should be assignable to `typeof e` in the absence of a concrete type for `T`. In contrast, the userland `Spread<T, U>` will not be assignable to `T` or `U`, nor would a hypothetical built-in type spread operator.

So anyway for concrete types, `Spread` is probably better, but for generics (or when at least one operand is generic) `T & U` is probably better, but there's no facility for overloading on generics.

**Comment by Benjamin-Dobell** [comment]
> it's desirable for Object.assign(T, U) for two type parameters T and U to be assignable to both T and U

That seems like a very peculiar thing to infer. I can understand why people may be expecting `Object.assign(T, ...)` to be assignable to `T`, but making assumptions about the source parameters seems particularly odd (they're almost always partials). Granted I don't at all doubt you when you say you've run into it.

Given that TypeScript wants to facilitate writing correct & maintainable software, I would ideally like to see changes that cause developers to disperse with these sorts of false assumptions. However, I totally understand what's proposed is a rather large change; and I'm certainly not expecting anyone to rashly make decisions. Nonetheless, I'm a bit sad to see this labelled as 'Working as intended' given the design goals of the language.

Anyway, I'll just keep my fingers crossed there'll be a change of heart at some point 🤞

**Comment by RyanCavanaugh** [comment]
Maybe "Best Available Compromise" is the better label
